ui/apps/platform/src/Containers/NetworkGraph/TopologyComponent.tsx

```diff
@@ -34,6 +34,60 @@ import {
 import { EdgeState } from './components/EdgeStateSelect';
 import { deploymentTabs } from './utils/deploymentUtils';
 import { NetworkScopeHierarchy } from './types/networkScopeHierarchy';
+import { EntityScope } from './simulation/NetworkPoliciesGenerationScope';
+
+/**
+ * Given the array of nodeModels returned by the network graph API and the user's
+ * current scope hierarchy, return the scope of entities that should be included
+ * in the network policy simulation.
+ */
```

Comment:
This is needed to get an accurate list of the deployments that will potentially have generated network policies. We need to use **both** the graph response from the server and the current selected scope for the following reasons:

- The user can select Namespace=['frontend', 'backend', 'operations'] as well as Deployment=['nginx'] from the scope dropdowns. This will include all deployments in any of those three namespaces that start with the string "nginx", and we can't know how many there will be based on this information alone.
- Similarly, the user could apply a general filter like "Deployment Label=app/nginx3" that could reduce the number of deployments that would be included in the scope above.

Given the server response, we will have a list of all deployments that match the applied scope and filters, as well as deployments that have an active flow to these deployments. I believe we could get the correct list only with this data, but checking if the individual deployments are included in the user's scope makes the logic of filtering out deployments that are only connected by "flow" faster and clearer.

Reply:
If this is a pure function, what are pro and con to extract it into a separate file in the same folder?

Reply:
Given the possibility in previous comment, it becomes more practical to make a tactical extraction of a pure helper function that given `scopeHierarchy` returns `granularity` instead of imperative style `let` assignment with case analysis to reassign.

Reply:
For these single use functions I usually keep them colocated in the component that uses them, although this one did feel like a bit of an odd case since the purpose of the function is essentially a data transformer from one component to the next. I can extract into a util and split the functions if it aids in readability.